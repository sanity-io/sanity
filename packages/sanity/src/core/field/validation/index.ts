import {SchemaType, ObjectSchemaType} from '@sanity/types'

/** @internal */
export interface FieldValueError {
  message: string
  value: unknown
}

/** @internal */
export function getValueError(value: unknown, schemaType: SchemaType): FieldValueError | undefined {
  const {jsonType} = schemaType
  const valueType = Array.isArray(value) ? 'array' : typeof value

  if (value === null || valueType === 'undefined') {
    return undefined
  }

  if (valueType !== jsonType) {
    return {message: `Value is ${valueType}, expected ${jsonType}`, value}
  }

  if (isObjectType(schemaType) && isObjectValue(value)) {
    for (const field of schemaType.fields) {
      const fieldError = getValueError(value[field.name], field.type as SchemaType)
      if (fieldError) {
        return fieldError
      }
    }
  }

  return undefined
}

function isObjectType(schemaType: SchemaType): schemaType is ObjectSchemaType {
  return schemaType.jsonType === 'object'
}

function isObjectValue(value: unknown): value is Record<string, unknown> {
  return value !== null && !Array.isArray(value) && typeof value === 'object'
}
