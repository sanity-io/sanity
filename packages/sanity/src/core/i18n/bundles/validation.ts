/* eslint sort-keys: "error" */
import {defineLocaleResourceBundle, defineLocalesResources} from '../helpers'
import {validationLocaleNamespace} from '../localeNamespaces'

/**
 * The string resources for validation.
 *
 * @internal
 */
const validationLocaleStrings = defineLocalesResources('validation', {
  /** Array must have exactly "$wantedLength" items, but has more/less */
  'array.exact-length': 'Must have exactly {{wantedLength}} items',
  /** Portable Text array must have exactly "$wantedLength" blocks, but has more/less */
  'array.exact-length_blocks': 'Must have exactly {{wantedLength}} blocks',
  /** Array item is a duplicate, but array wants only unique items */
  'array.item-duplicate': "Can't be a duplicate",
  /** Array has more than the maximum of "$maxLength" items */
  'array.maximum-length': 'Must have at most {{maxLength}} items',
  /** Portable Text array has more than the maximum of "$maxLength" items */
  'array.maximum-length_blocks': 'Must have at most {{maxLength}} blocks',
  /** Array has less than the minimum of "$minLength" items */
  'array.minimum-length': 'Must have at least {{minLength}} items',
  /** Portable Text array has less than the minimum of "$minLength" blocks */
  'array.minimum-length_blocks': 'Must have at least {{minLength}} blocks',
  /** Date is not valid or not in the correct format (ISO-8601) */
  'date.invalid-format': 'Must be a valid ISO-8601 formatted date string',
  /** Date is later than the given maximum date "$maxDate" */
  'date.maximum': 'Must be at or before {{maxDate}}',
  /** Date is earlier than the given minimum date "$minDate" */
  'date.minimum': 'Must be at or after {{minDate}}',
  /** A value of incorrect type is found, eg found `number` instead of `string` */
  'generic.incorrect-type': 'Expected type "{{expectedType}}", got "{{actualType}}"',
  /** Value is not one of the values specifically allowed */
  'generic.not-allowed': 'Value did not match any allowed values',
  /** Value "$givenValue" is not one of the values specifically allowed */
  'generic.not-allowed_hint': 'Value "{{hint}}" did not match any allowed values',
  /** A value is expected, but none is provided */
  'generic.required': 'Required',
  /** Number is less than the given minimum threshold value "$threshold" */
  'number.greater-than': 'Must be greater than {{threshold}}',
  /** Number is greater than the given maximum threshold value "$threshold" */
  'number.less-than': 'Must be less than {{threshold}}',
  /** Number is higher than the given maximum value "$maxNumber" */
  'number.maximum': 'Must be lower than or equal to {{maxNumber}}',
  /** Number has more precision (decimals) than the allowed "$limit" */
  'number.maximum-precision': 'Max precision is {{limit}}',
  /** Number is lower than the given minimum value "$minNumber" */
  'number.minimum': 'Must be greater than or equal to {{minNumber}}',
  /** Number is not an integer ("whole number") */
  'number.non-integer': 'Must be an integer',
  /** Object is missing a reference to an asset document in its `asset` field */
  'object.asset-required': 'Asset is required',
  /** Object is missing a reference to a file asset document in its `asset` field */
  'object.asset-required_file': 'File is required',
  /** Object is missing a reference to an image asset document in its `asset` field */
  'object.asset-required_image': 'Image is required',
  /** Object is not a reference to a document (eg `{_ref: 'documentId'}`) */
  'object.not-reference': 'Must be a reference to a document',
  /** Object references a document which is not published */
  'object.reference-not-published': 'Referenced document must be published',
  /** Accessibility label for closing the validation panel */
  'panel.close-button-aria-label': 'Close validation',
  /** Message shown when the validation panel is opened but there are no errors/warnings */
  'panel.no-errors-message': 'No validation errors',
  /** Title for the actual "Validation" panel/feature */
  'panel.title': 'Validation',
  /** Slug is an object, but is missing a `current` string property */
  'slug.missing-current': 'Slug must have a value',
  /** Slug is not an object (eg `{current: 'some-slug'}`) */
  'slug.not-object': 'Slug must be an object',
  /** Slug is already in use somewhere else, but needs to be unique */
  'slug.not-unique': 'Slug is already in use',
  /** String is not a valid email address */
  'string.email': 'Must be a valid email address',
  /** String has a different character length than the exact number "$wantedLength" */
  'string.exact-length': 'Must be exactly {{wantedLength}} characters long',
  /** String contains characters that are not in lowercase  */
  'string.lowercase': 'Must be all lowercase characters',
  /** String is longer than the limit of "$maxLength" characters */
  'string.maximum-length': 'Must be at most {{maxLength}} characters long',
  /** String is shorter than the limit of "$minLength" characters */
  'string.minimum-length': 'Must be at least {{minLength}} characters long',
  /** String does not match the given regular expression, but should */
  'string.regex-does-not-match': 'Does not match "{{name}}"-pattern',
  /** String matches the given regular expression, but should not */
  'string.regex-match': 'Should not match "{{name}}"-pattern',
  /** String contains characters that are not in uppercase */
  'string.uppercase': 'Must be all uppercase characters',
  /** String contains a protocol/scheme that is not allowed, eg (`ftp`, `mailto`â€¦) */
  'string.url.disallowed-scheme': 'Does not match allowed protocols/schemes',
  /** String contains a URL with a username or password specified before the host */
  'string.url.includes-credentials': 'Username/password not allowed',
  /** String is not a valid URL */
  'string.url.invalid': 'Not a valid URL',
  /** String is not an absolute URL (eg it is missing a protocol/host) */
  'string.url.not-absolute': 'Relative URLs are not allowed',
  /** String is not a relative URL (eg it contains a protocol/host) */
  'string.url.not-relative': 'Only relative URLs are allowed',
} as const)

/**
 * The i18n resource keys for the validation.
 *
 * @alpha
 * @hidden
 */
export type ValidationLocaleResourceKeys = keyof typeof validationLocaleStrings

/**
 * Locale resources for the validation namespace, eg US English locale resources.
 *
 * @beta
 * @hidden
 */
export const validationLocaleResources = defineLocaleResourceBundle({
  locale: 'en-US',
  namespace: validationLocaleNamespace,
  resources: validationLocaleStrings,
})
