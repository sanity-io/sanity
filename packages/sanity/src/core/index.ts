export * from './FIXME'
export * from './changeIndicators'
export * from './components'
export * from './components/collapseMenu'
export * from './components/scroll'
export * from './config'
export * from './environment'
export * from './field'
export * from './form'
export * from './hooks'
export * from './presence'
export * from './preview'
export * from './schema'
export * from './store'
export * from './studio'
export * from './studioClient'
export * from './templates'
export * from './theme'
export * from './user-color'
export * from './util'
export * from './version'
export {createSearchQuery} from './search'
export type {
  SearchableType,
  SearchTerms,
  SearchOptions,
  SearchSort,
  WeightedSearchOptions,
} from './search'
