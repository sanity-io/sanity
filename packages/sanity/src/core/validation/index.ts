export {Rule} from './Rule'
export {validateDocument} from './validateDocument'
export {inferFromSchema} from './inferFromSchema'
export {inferFromSchemaType} from './inferFromSchemaType'
export {validateDocumentObservable} from './validateDocument'
