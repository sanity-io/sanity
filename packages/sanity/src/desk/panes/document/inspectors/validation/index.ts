import {CheckmarkCircleIcon, ErrorOutlineIcon, WarningOutlineIcon} from '@sanity/icons'
import {useMemo} from 'react'
import {VALIDATION_INSPECTOR_NAME} from '../../constants'
import {ValidationInspector} from './ValidationInspector'
import {
  DocumentInspector,
  DocumentInspectorMenuItem,
  DocumentInspectorUseMenuItemProps,
  FormNodeValidation,
  isValidationError,
  isValidationWarning,
  useValidationStatus,
} from 'sanity'

function useMenuItem(props: DocumentInspectorUseMenuItemProps): DocumentInspectorMenuItem {
  const {documentId, documentType} = props
  const {validation: validationMarkers} = useValidationStatus(documentId, documentType)

  const validation: FormNodeValidation[] = useMemo(
    () =>
      validationMarkers.map((item) => ({
        level: item.level,
        message: item.item.message,
        path: item.path,
      })),
    [validationMarkers],
  )

  const hasErrors = validation.some(isValidationError)
  const hasWarnings = validation.some(isValidationWarning)

  const icon = useMemo(() => {
    if (hasErrors) return ErrorOutlineIcon
    if (hasWarnings) return WarningOutlineIcon
    return CheckmarkCircleIcon
  }, [hasErrors, hasWarnings])

  const tone = useMemo(() => {
    if (hasErrors) return 'critical'
    if (hasWarnings) return 'caution'
    return 'positive'
  }, [hasErrors, hasWarnings])

  return {
    hidden: validation.length === 0,
    icon,
    title: 'Validation',
    tone,
    showAsAction: true,
  }
}

export const validationInspector: DocumentInspector = {
  name: VALIDATION_INSPECTOR_NAME,
  component: ValidationInspector,
  useMenuItem,
}
