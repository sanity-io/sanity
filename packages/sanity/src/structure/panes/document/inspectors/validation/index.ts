import {CheckmarkCircleIcon, ErrorOutlineIcon, WarningOutlineIcon} from '@sanity/icons'
import {useMemo} from 'react'
import {
  type DocumentInspector,
  type DocumentInspectorMenuItem,
  type DocumentInspectorUseMenuItemProps,
  type FormNodeValidation,
  isGoingToUnpublish,
  isValidationError,
  isValidationWarning,
  usePerspective,
  useTranslation,
  useValidationStatus,
} from 'sanity'

import {VALIDATION_INSPECTOR_NAME} from '../../constants'
import {useDocumentPane} from '../../useDocumentPane'
import {ValidationInspector} from './ValidationInspector'

function useMenuItem(props: DocumentInspectorUseMenuItemProps): DocumentInspectorMenuItem {
  const {documentId, documentType} = props
  const {t} = useTranslation('validation')
  const {selectedReleaseId} = usePerspective()
  const {validation: validationMarkers} = useValidationStatus(
    documentId,
    documentType,
    selectedReleaseId,
  )
  const {value} = useDocumentPane()

  const validation: FormNodeValidation[] = useMemo(
    () =>
      validationMarkers.map((item) => ({
        level: item.level,
        message: item.message,
        path: item.path,
      })),
    [validationMarkers],
  )

  const hasErrors = validation.some(isValidationError)
  const hasWarnings = validation.some(isValidationWarning)
  const isDocumentGoingToUnpublish = isGoingToUnpublish(value)

  const icon = useMemo(() => {
    if (hasErrors) return ErrorOutlineIcon
    if (hasWarnings) return WarningOutlineIcon
    return CheckmarkCircleIcon
  }, [hasErrors, hasWarnings])

  const tone = useMemo(() => {
    if (hasErrors) return 'critical'
    if (hasWarnings) return 'caution'
    return 'positive'
  }, [hasErrors, hasWarnings])

  return {
    hidden: validation.length === 0 || isDocumentGoingToUnpublish,
    icon,
    title: t('panel.title'),
    tone,
    showAsAction: true,
  }
}

export const validationInspector: DocumentInspector = {
  name: VALIDATION_INSPECTOR_NAME,
  component: ValidationInspector,
  useMenuItem,
}
