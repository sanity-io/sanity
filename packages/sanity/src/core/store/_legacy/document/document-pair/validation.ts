import type {SanityClient} from '@sanity/client'
import {
  asyncScheduler,
  combineLatest,
  concat,
  defer,
  from,
  lastValueFrom,
  Observable,
  of,
  timer,
} from 'rxjs'
import {
  distinct,
  distinctUntilChanged,
  first,
  groupBy,
  map,
  mergeMap,
  publishReplay,
  refCount,
  scan,
  share,
  shareReplay,
  skip,
  throttleTime,
} from 'rxjs/operators'
import {isReference, Schema, ValidationContext, ValidationMarker} from '@sanity/types'
import {reduce as reduceJSON} from 'json-reduce'
import shallowEquals from 'shallow-equals'
import {omit} from 'lodash'
import {exhaustMapWithTrailing} from 'rxjs-exhaustmap-with-trailing'
import {validateDocumentObservable} from '../../../../validation'
import {SourceClientOptions} from '../../../../config'
import {memoize} from '../utils/createMemoizer'
import {IdPair} from '../types'
import {DraftsModelDocumentAvailability} from '../../../../preview'
import {editState} from './editState'

/**
 * @hidden
 * @beta */
export interface ValidationStatus {
  isValidating: boolean
  validation: ValidationMarker[]
  revision?: string
}

const INITIAL_VALIDATION_STATUS: ValidationStatus = {
  isValidating: true,
  validation: [],
}

function findReferenceIds(obj: any): Set<string> {
  return reduceJSON(
    obj,
    (acc, node) => {
      if (isReference(node)) {
        acc.add(node._ref)
      }
      return acc
    },
    new Set<string>(),
  )
}

const EMPTY_VALIDATION: ValidationMarker[] = []

type GetDocumentExists = NonNullable<ValidationContext['getDocumentExists']>

type ObserveDocumentPairAvailability = (id: string) => Observable<DraftsModelDocumentAvailability>

const listenDocumentExists = (
  observeDocumentAvailability: ObserveDocumentPairAvailability,
  id: string,
): Observable<boolean> =>
  observeDocumentAvailability(id).pipe(map(({published}) => published.available))

// throttle delay for document updates (i.e. time between responding to changes in the current document)
const DOC_UPDATE_DELAY = 200

// throttle delay for referenced document updates (i.e. time between responding to changes in referenced documents)
const REF_UPDATE_DELAY = 1000

/** @internal */
export const validation = memoize(
  (
    ctx: {
      client: SanityClient
      getClient: (options: SourceClientOptions) => SanityClient
      observeDocumentPairAvailability: ObserveDocumentPairAvailability
      schema: Schema
    },
    {draftId, publishedId}: IdPair,
    typeName: string,
  ): Observable<ValidationStatus> => {
    const document$ = editState(ctx, {draftId, publishedId}, typeName).pipe(
      map(({draft, published}) => draft || published),
      throttleTime(DOC_UPDATE_DELAY, asyncScheduler, {trailing: true}),
      distinctUntilChanged((prev, next) => {
        if (prev?._rev === next?._rev) {
          return true
        }
        // _rev and _updatedAt may change without other fields changing (due to a limitation in mutator)
        // so only pass on documents if _other_ attributes changes
        return shallowEquals(omit(prev, '_rev', '_updatedAt'), omit(next, '_rev', '_updatedAt'))
      }),
      share(),
    )

    const referenceIds$ = document$.pipe(
      map((document) => findReferenceIds(document)),
      mergeMap((ids) => from(ids)),
    )

    // Note: we only use this to trigger a re-run of validation when a referenced document is published/unpublished
    const referenceExistence$ = referenceIds$.pipe(
      groupBy(
        (id) => id,
        undefined,
        () => timer(1000 * 60 * 30),
      ),
      mergeMap((id$) =>
        id$.pipe(
          distinct(),
          mergeMap((id) =>
            listenDocumentExists(ctx.observeDocumentPairAvailability, id).pipe(
              map(
                // eslint-disable-next-line max-nested-callbacks
                (result) => [id, result] as const,
              ),
            ),
          ),
        ),
      ),
      scan((acc: Record<string, boolean>, [id, result]): Record<string, boolean> => {
        if (Boolean(acc[id]) === result) {
          return acc
        }
        return result ? {...acc, [id]: result} : omit(acc, id)
      }, {}),
      distinctUntilChanged(shallowEquals),
      shareReplay({refCount: true, bufferSize: 1}),
    )

    // Provided to individual validation functions to support using existence of a weakly referenced document
    // as part of the validation rule (used by references in place)
    const getDocumentExists: GetDocumentExists = ({id}) =>
      lastValueFrom(
        referenceExistence$.pipe(
          // If the id is not present as key in the `referenceExistence` map it means it's existence status
          // isn't yet loaded, so we want to wait until it is
          first((referenceExistence) => id in referenceExistence),
          map((referenceExistence) => referenceExistence[id]),
        ),
      )

    const referenceDocumentUpdates$ = referenceExistence$.pipe(
      // we'll skip the first emission since the document already gets an initial validation pass
      // we're only interested in updates in referenced documents after that
      skip(1),
      throttleTime(REF_UPDATE_DELAY, asyncScheduler, {leading: true, trailing: true}),
    )

    return combineLatest([document$, concat(of(null), referenceDocumentUpdates$)]).pipe(
      map(([document]) => document),
      exhaustMapWithTrailing((document) => {
        return defer(() => {
          if (!document?._type) {
            return of({validation: EMPTY_VALIDATION, isValidating: false})
          }
          return concat(
            of({isValidating: true, revision: document._rev}),
            validateDocumentObservable(ctx.getClient, document, ctx.schema, {
              getDocumentExists,
            }).pipe(
              map((validationMarkers) => ({validation: validationMarkers, isValidating: false})),
            ),
          )
        })
      }),
      scan((acc, next) => ({...acc, ...next}), INITIAL_VALIDATION_STATUS),
      publishReplay(1),
      refCount(),
    )
  },
  (ctx, idPair, typeName) => {
    const config = ctx.client.config()

    return `${config.dataset ?? ''}-${config.projectId ?? ''}-${idPair.publishedId}-${typeName}`
  },
)
