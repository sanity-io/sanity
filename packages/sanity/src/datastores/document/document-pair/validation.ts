import type {SanityClient} from '@sanity/client'
import {concat, of, combineLatest, defer, from, Observable} from 'rxjs'
import {
  map,
  scan,
  switchMap,
  mergeMap,
  publishReplay,
  refCount,
  share,
  distinctUntilChanged,
  debounceTime,
  first,
} from 'rxjs/operators'
import {validateDocument} from '@sanity/validation'
import {ValidationMarker, ValidationContext, isReference, Schema} from '@sanity/types'
import reduceJSON from 'json-reduce'
import {SourceClientOptions} from '../../../config'
import {memoize} from '../utils/createMemoizer'
import {IdPair} from '../types'
import {HistoryStore} from '../../history'
import {DocumentPreviewStore} from '../../../preview'
import {editState} from './editState'

export interface ValidationStatus {
  isValidating: boolean
  validation: ValidationMarker[]
}

const INITIAL_VALIDATION_STATUS: ValidationStatus = {isValidating: true, validation: []}

function findReferenceIds(obj: any): Set<string> {
  return reduceJSON(
    obj,
    (acc, node) => {
      if (isReference(node)) {
        acc.add(node._ref)
      }
      return acc
    },
    new Set<string>()
  )
}

type GetDocumentExists = NonNullable<ValidationContext['getDocumentExists']>

const listenDocumentExists = (
  previewStore: DocumentPreviewStore,
  id: string
): Observable<boolean> =>
  previewStore
    .unstable_observeDocumentPairAvailability({_type: 'reference', _ref: id})
    .pipe(map(({published}) => published.available))

export const validation = memoize(
  (
    ctx: {
      client: SanityClient
      getClient: (options: SourceClientOptions) => SanityClient
      documentPreviewStore: DocumentPreviewStore
      historyStore: HistoryStore
      schema: Schema
    },
    {draftId, publishedId}: IdPair,
    typeName: string
  ): Observable<ValidationStatus> => {
    const getDocumentExists: GetDocumentExists = ({id}) =>
      listenDocumentExists(ctx.documentPreviewStore, id).pipe(first()).toPromise()

    const document$ = editState(ctx, {draftId, publishedId}, typeName).pipe(
      map(({draft, published}) => draft || published),
      // this debounce is needed for performance. it prevents the validation
      // from being run on every keypress
      debounceTime(300),
      share()
    )

    const referenceIds$ = document$.pipe(
      map((document) => findReferenceIds(document)),
      distinctUntilChanged((curr, next) => {
        if (curr.size !== next.size) return false
        for (const item of curr) {
          if (!next.has(item)) return false
        }
        return true
      })
    )

    const referencedDocumentUpdate$ = referenceIds$.pipe(
      switchMap((idSet) =>
        from(idSet).pipe(
          mergeMap((id) =>
            listenDocumentExists(ctx.documentPreviewStore, id).pipe(
              map(
                // eslint-disable-next-line max-nested-callbacks
                (result) => [id, result] as const
              )
            )
          ),
          // the `debounceTime` in the next stream removes multiple emissions
          // caused by this scan
          scan(
            (acc, [id, result]) => ({
              ...acc,
              [id]: result,
            }),
            {} as Record<string, boolean>
          )
        )
      ),
      distinctUntilChanged((curr, next) => {
        const currKeys = Object.keys(curr)
        const nextKeys = Object.keys(next)
        if (currKeys.length !== nextKeys.length) return false
        for (const key of currKeys) {
          if (curr[key] !== next[key]) return false
        }
        return true
      })
    )

    return combineLatest([
      // from document edits
      document$,
      // and from document dependency events
      concat(
        // note: that the `referencedDocumentUpdate$` may not pre-emit any
        // events (unlike `editState` which includes `publishReplay(1)`), so
        // we `concat` the stream with an empty emission so `combineLatest` will
        // emit as soon as `editState` emits
        //
        // > Be aware that `combineLatest` will not emit an initial value until
        // > each observable emits at least one value.
        // https://www.learnrxjs.io/learn-rxjs/operators/combination/combinelatest#why-use-combinelatest
        of(null),
        referencedDocumentUpdate$
      ).pipe(
        // don't remove, see `debounceTime` comment above
        debounceTime(50)
      ),
    ]).pipe(
      map(([document]) => document),
      switchMap((document) =>
        concat(
          of({isValidating: true}),
          defer(async () => {
            if (!document?._type) {
              return {validation: [], isValidating: false}
            }

            // TODO: consider cancellation eventually
            const markers = await validateDocument(ctx.getClient, document, ctx.schema, {
              getDocumentExists,
            })

            return {validation: markers, isValidating: false}
          })
        )
      ),
      scan((acc, next) => ({...acc, ...next}), INITIAL_VALIDATION_STATUS),
      publishReplay(1),
      refCount()
    )
  },
  (_ctx, idPair) => idPair.publishedId
)
