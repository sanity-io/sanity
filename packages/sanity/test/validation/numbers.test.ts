import {Rule} from '../../src/core/validation'

const context: any = {client: {}}

describe('number', () => {
  test('required constraint', async () => {
    const rule = Rule.number().required()
    await expect(rule.validate(undefined, context)).resolves.toMatchSnapshot('required: undefined')
    await expect(rule.validate(13, context)).resolves.toMatchSnapshot('required: valid')
  })

  test('min constraint', async () => {
    const rule = Rule.number().min(10)
    await expect(rule.validate(3, context)).resolves.toMatchSnapshot('min: too low')
    await expect(rule.validate(10, context)).resolves.toMatchSnapshot('min: valid, at limit')
    await expect(rule.validate(20, context)).resolves.toMatchSnapshot('min: valid')
  })

  test('greater than constraint', async () => {
    const rule = Rule.number().greaterThan(10)
    await expect(rule.validate(3, context)).resolves.toMatchSnapshot('gt: too low')
    await expect(rule.validate(10, context)).resolves.toMatchSnapshot('gt: too low, at limit')
    await expect(rule.validate(20, context)).resolves.toMatchSnapshot('gt: valid')
  })

  test('max constraint', async () => {
    const rule = Rule.number().max(10)
    await expect(rule.validate(20, context)).resolves.toMatchSnapshot('max: too large')
    await expect(rule.validate(10, context)).resolves.toMatchSnapshot('max: valid, at limit')
    await expect(rule.validate(5, context)).resolves.toMatchSnapshot('max: valid')
  })

  test('less than constraint', async () => {
    const rule = Rule.number().lessThan(10)
    await expect(rule.validate(20, context)).resolves.toMatchSnapshot('lt: too high')
    await expect(rule.validate(10, context)).resolves.toMatchSnapshot('lt: too high, at limit')
    await expect(rule.validate(3, context)).resolves.toMatchSnapshot('lt: valid')
  })

  test('integer constraint', async () => {
    const rule = Rule.number().integer()
    await expect(rule.validate(31.14, context)).resolves.toMatchSnapshot('integer: invalid (float)')
    await expect(rule.validate(31, context)).resolves.toMatchSnapshot('integer: valid')
  })

  test('precision constraint', async () => {
    const rule = Rule.number().precision(3)
    await expect(rule.validate(Math.PI, context)).resolves.toMatchSnapshot(
      'precision: invalid (pi)'
    )
    await expect(rule.validate(31.133, context)).resolves.toMatchSnapshot(
      'precision: valid (at limit)'
    )
    await expect(rule.validate(31.3, context)).resolves.toMatchSnapshot(
      'precision: valid (below limit)'
    )
  })

  test('positive constraint', async () => {
    const rule = Rule.number().positive()
    await expect(rule.validate(-31.14, context)).resolves.toMatchSnapshot('positive: invalid')
    await expect(rule.validate(0, context)).resolves.toMatchSnapshot('positive: valid (zero)')
    await expect(rule.validate(13, context)).resolves.toMatchSnapshot('positive: valid')
  })

  test('negative constraint', async () => {
    const rule = Rule.number().negative()
    await expect(rule.validate(31.14, context)).resolves.toMatchSnapshot('negative: invalid')
    await expect(rule.validate(0, context)).resolves.toMatchSnapshot('negative: invalid (zero)')
    await expect(rule.validate(-13, context)).resolves.toMatchSnapshot('negative: valid')
  })
})
