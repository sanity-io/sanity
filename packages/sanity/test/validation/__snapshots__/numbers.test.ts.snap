// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`number greater than constraint: gt: too low 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Must be greater than 10",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
  },
]
`;

exports[`number greater than constraint: gt: too low, at limit 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Must be greater than 10",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
  },
]
`;

exports[`number greater than constraint: gt: valid 1`] = `Array []`;

exports[`number integer constraint: integer: invalid (float) 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Must be an integer",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
  },
]
`;

exports[`number integer constraint: integer: valid 1`] = `Array []`;

exports[`number less than constraint: lt: too high 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Must be less than 10",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
  },
]
`;

exports[`number less than constraint: lt: too high, at limit 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Must be less than 10",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
  },
]
`;

exports[`number less than constraint: lt: valid 1`] = `Array []`;

exports[`number max constraint: max: too large 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Must be less than or equal 10",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
  },
]
`;

exports[`number max constraint: max: valid 1`] = `Array []`;

exports[`number max constraint: max: valid, at limit 1`] = `Array []`;

exports[`number min constraint: min: too low 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Must be greater than or equal 10",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
  },
]
`;

exports[`number min constraint: min: valid 1`] = `Array []`;

exports[`number min constraint: min: valid, at limit 1`] = `Array []`;

exports[`number negative constraint: negative: invalid (zero) 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Must be less than 0",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
  },
]
`;

exports[`number negative constraint: negative: invalid 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Must be less than 0",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
  },
]
`;

exports[`number negative constraint: negative: valid 1`] = `Array []`;

exports[`number positive constraint: positive: invalid 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Must be greater than or equal 0",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
  },
]
`;

exports[`number positive constraint: positive: valid (zero) 1`] = `Array []`;

exports[`number positive constraint: positive: valid 1`] = `Array []`;

exports[`number precision constraint: precision: invalid (pi) 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Max precision is 3",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
  },
]
`;

exports[`number precision constraint: precision: valid (at limit) 1`] = `Array []`;

exports[`number precision constraint: precision: valid (below limit) 1`] = `Array []`;

exports[`number required constraint: required: undefined 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Value is required",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
  },
]
`;

exports[`number required constraint: required: valid 1`] = `Array []`;
